src/jobserver-posix.cc

```diff
@@ -0,0 +1,170 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "jobserver.h"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <cassert>
+#include <cstring>
+
+#include "util.h"
+
+Jobserver::Jobserver() {
+  assert(!Enabled());
+
+  // Return early if no makeflags are passed in the environment.
+  char* makeflags = std::getenv("MAKEFLAGS");
+  if (makeflags == nullptr || strlen(makeflags) == 0)
+    return;
+
+  std::string::size_type flag_char_ = 0;
+  std::string flag_;
+  std::vector<std::string> flags_;
+
+  // Tokenize string to characters in flag_, then words in flags_.
+  while (flag_char_ < strlen(makeflags)) {
+    while (flag_char_ < strlen(makeflags) &&
+           !isblank(static_cast<unsigned char>(makeflags[flag_char_]))) {
+      flag_.push_back(static_cast<unsigned char>(makeflags[flag_char_]));
+      flag_char_++;
+    }
+
+    if (!flag_.empty())
+      flags_.push_back(flag_);
+
+    flag_.clear();
+    flag_char_++;
+  }
+
+  // --jobserver-auth=<val>
+  for (size_t n = 0; n < flags_.size(); n++)
+    if (flags_[n].find(AUTH_KEY) == 0)
+      flag_ = flags_[n].substr(strlen(AUTH_KEY));
+
+  // --jobserver-fds=<val>
+  if (flag_.empty())
+    for (size_t n = 0; n < flags_.size(); n++)
+      if (flags_[n].find(FDS_KEY) == 0)
+        flag_ = flags_[n].substr(strlen(FDS_KEY));
```

Comment:
These two successive loops means that `--jobserver-fds=4,5 --jobserver-auth=pipe:<name>` will not use the pipe method, even though it should. It would be nicer if this was handled properly.

Also maybe add a comment explaning what `--jobserver-fds`  is, because it is not on the official GNU Make documentation, since it was supposed to be used internally until GNU Make 4.2 renamed it to `--jobserver-auth`.

https://lists.gnu.org/archive/html/info-gnu/2016-05/msg00013.html

Reply:
> These two successive loops means that `--jobserver-fds=4,5 --jobserver-auth=pipe:<name>` will not use the pipe method, even though it should. It would be nicer if this was handled properly.

It is handled properly 👍🏼 

This is one of the first changes I made to #2450 so it has the most testing behind it so far.

First, a match to `--jobserver-auth` is searched and if found, `flag` is set. Then, only if `flag` is empty, search for the deprecated name. Then (again), only if `flag` is empty, look for the possibility of Make being used with `-j 1`.

> Also maybe add a comment explaning what `--jobserver-fds` is, because it is not on the official GNU Make documentation, since it was supposed to be used internally until GNU Make 4.2 renamed it to `--jobserver-auth`.

This is quickly mentioned in the comments of the header but I can expand on that if you want.